### PR TITLE
Fix sharing both an image and a text to DC

### PR DESCRIPTION
**Pull request by Hocuri**
This bug was re-introduced in https://github.com/deltachat/deltachat-android/pull/1475/.

Steps: Share media+text from another app to DC.

@cyBerta for the next time: could you please ping me when you revert one of my commits?
@r10s please wait until @cyBerta commented, maybe it is possible to prevent adding another bug again this time.

Concerning this SettableFuture: Turns out it is only used to determine whether to go to lastseen (or not) but this does not work anyway: #1528.

### CODE REVIEW

**Comment by Hocuri** [comment]
@r10s if `fragment.moveToLastSeen();` is doing something that currently works, then this pr might break it. Because I am not sure if initializeDraft() still always returns true or false correctly and in https://github.com/deltachat/deltachat-android/blob/46a14b954cb2e0a23282776dcc9abb0d0215b6dc/src/org/thoughtcrime/securesms/ConversationActivity.java#L238 it depends on the returned value whether moveToLastSeen() is invoked or not.

However, I still do not understand this feature and why we would only call moveToLastSeen() if the share draft was initialized.

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 1]
```diff
@@ -725,26 +725,28 @@ public void onSuccess(Boolean result) {
    */
   private ListenableFuture<Boolean> initializeDraft() {
 
-    final SettableFuture<Boolean> result = new SettableFuture<>();
+    SettableFuture<Boolean> result = new SettableFuture<>();
 
-    final String    draftText      = getIntent().getStringExtra(TEXT_EXTRA);
```
Comment (cyBerta) [inline src/org/thoughtcrime/securesms/ConversationActivity.java:730]: This reintroduces #1472.

Comment (cyBerta) [reply src/org/thoughtcrime/securesms/ConversationActivity.java:730]: I'll add more comments later

**Comment by cyBerta** [comment]
Thanks for your efforts @Hocuri 
since it's already late, I'll continue tomorrow. I didn't completely understand the fix yet, but tomorrow with fresh energy I will hopefully.

**Comment by cyBerta** [comment]
in general I think this part is too complex and hard to follow. Several paths lead to the initialization part, initialization from an intent (e.g. through sendto intent) and drafts from the database are handled here. these cases can have different passed values and with regard to that the implications of changes like that https://github.com/deltachat/deltachat-android/pull/1529/files#diff-8b6badc7254f2d75634d87a85bc9cd31R744 are not easy to understand. That's why I think splitting it the method up into several for different cases might be an improvement.

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 2]
```diff
@@ -725,26 +725,28 @@ public void onSuccess(Boolean result) {
...
 
-    if (draftText == null && draftMedia == null && draftMediaType == null) {
-      return initializeDraftFromDatabase();
+    if (draftText == null || draftMedia == null || draftMediaType == null) {
+      result = initializeDraftFromDatabase();
     } else {
-      updateToggleButtonState();
       result.set(false);
     }
+    updateToggleButtonState();
```
Comment (cyBerta) [inline src/org/thoughtcrime/securesms/ConversationActivity.java:749]: updateToggleButtonState() is called in initializeDraftFromDatabase(), no need to call it twice

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 3]
```diff
@@ -725,26 +725,28 @@ public void onSuccess(Boolean result) {
...
-      return setMedia(draftMedia, draftMediaType);
+      result = setMedia(draftMedia, draftMediaType);
     }
 
-    if (draftText == null && draftMedia == null && draftMediaType == null) {
-      return initializeDraftFromDatabase();
+    if (draftText == null || draftMedia == null || draftMediaType == null) {
+      result = initializeDraftFromDatabase();
     } else {
-      updateToggleButtonState();
```
Comment (cyBerta) [inline src/org/thoughtcrime/securesms/ConversationActivity.java:745]: only in this case we need to make sure updateToggleButton gets called

**Comment by cyBerta** [comment]
> Because I am not sure if initializeDraft() still always returns true or false correctly

There's another usage that depends on the correct return value: 
https://github.com/deltachat/deltachat-android/blob/d10891311fa706827f3ec50263c1298a7537332b/src/org/thoughtcrime/securesms/ConversationActivity.java#L711

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 4]
```diff
@@ -725,26 +725,28 @@ public void onSuccess(Boolean result) {
    */
   private ListenableFuture<Boolean> initializeDraft() {
 
-    final SettableFuture<Boolean> result = new SettableFuture<>();
+    SettableFuture<Boolean> result = new SettableFuture<>();
 
-    final String    draftText      = getIntent().getStringExtra(TEXT_EXTRA);
-    final Uri       draftMedia     = getIntent().getData();
```
Comment (cyBerta) [inline src/org/thoughtcrime/securesms/ConversationActivity.java:731]: getIntent().getData() gets set here https://github.com/deltachat/deltachat-android/blob/d10891311fa706827f3ec50263c1298a7537332b/src/org/thoughtcrime/securesms/NewConversationActivity.java#L171, if a sendto link was clicked. 
Replacing it by `RelayUtil.getSharedUris(this);` doesn't work in that case, because the latter will return an empty list.

**Comment by cyBerta** [comment]
@Hocuri which app did you use to share both text and images at the same time?

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 5]
```diff
@@ -756,7 +758,7 @@ private void initializeEnabledCheck() {
     attachButton.setEnabled(enabled);
   }
 
-  private ListenableFuture<Boolean> initializeDraftFromDatabase() {
+  private SettableFuture<Boolean> initializeDraftFromDatabase() {
```
[the remark this answers is not in the document]
Comment (Hocuri) [reply src/org/thoughtcrime/securesms/ConversationActivity.java:761]: It does not compile otherwise, changing this was the quickest way I saw.

**Comment by Hocuri** [comment]
> There's another usage that depends on the correct return value: 
https://github.com/deltachat/deltachat-android/blob/d10891311fa706827f3ec50263c1298a7537332b/src/org/thoughtcrime/securesms/ConversationActivity.java#L711

But here the returned boolean is not checked (at least if I understood correctly what SettableFuture does)

> @Hocuri which app did you use to share both text and images at the same time?

WhatsApp, should be possible with Telegram or other messengers too.

> in general I think this part is too complex and hard to follow.

Yes, maybe I should rearrange the code a little more, I just hope that I won't introduce a new bug again. Maybe I should make an integration test for this.

File: src/org/thoughtcrime/securesms/ConversationActivity.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/ConversationActivity.java:761]: If you split initializeDraft into initializeDraftFromIntent() (`mailto:`) and initializeDraftFromDatabase() you shouldn't need that change anymore.

**Comment by cyBerta** [comment]
> But here the returned boolean is not checked (at least if I understood correctly what SettableFuture does)

https://github.com/deltachat/deltachat-android/blob/d10891311fa706827f3ec50263c1298a7537332b/src/org/thoughtcrime/securesms/ConversationActivity.java#L1166

the returned boolean is saved in `isShareDraftInitialized` and used to handle the activity backstack correctly.
